### PR TITLE
initial orm error refactor

### PR DESCRIPTION
**Pull request by ruseinov**
Work on https://github.com/iov-one/weave/issues/309
As requested an example PR with sample refactoring, it touches several extensions as orm is used in a lot of them. Orm errors are fully refactored and mostly reduced to common errors. 

Tests haven't been refactored as of yet and still using stdlib error package. 
Let's see if this makes sense and continue if it does.

### CODE REVIEW

**Comment by ruseinov** [comment]
Actually tests don't look like they need to be refactored as they were not using internal errors anyway. 
@ethanfrey feel free to review this to see if we can move on to next packages.

File: errors/errors.go [hunk 1]
```diff
@@ -5,8 +5,9 @@ import (
 
 	"github.com/pkg/errors"
 )
-
+// Global error registry, codes 1-99 are reserved for global errors, 0 is reserved for non-errors
```
Comment (alpe) [inline errors/errors.go]: req: 0 is reserved for internal errors where the description string is not exposed to any client

File: errors/common.go [hunk 2]
```diff
@@ -209,3 +209,7 @@ func ErrModifyChainID() error {
 func IsModifyChainIDErr(err error) bool {
 	return IsSameError(errModifyChainID, err)
 }
+
+func WithType(err error, obj interface{}) error {
+	return Wrap(err, fmt.Sprintf("%T", obj))
```
Comment (alpe) [inline errors/common.go]: pp: good idea to extract common functionality.  Maybe you could rename this to `WithTypeDescr`

File: errors/errors.go (hunk 1 above)
Comment (ruseinov) [reply errors/errors.go]: @alpe We had to change this as discussed in the initial error proposal PR fix with @ethanfrey, here's the link with a detailed explanation: https://github.com/iov-one/weave/pull/308#discussion_r255423652 . I know it is confusing, but it is what it is. Basically it has to do with 0 being a success abci code.

File: errors/common.go (hunk 2 above)
Comment (ruseinov) [reply errors/common.go]: @alpe good point, not sure that I like the verbosity of it though as it's kind of straightforward that type here refers to description, basically we append the type name to a parent error.

I propose to keep it the same for now, we can always refactor in one operation later if it still feels confusing.

**Comment by ruseinov** [comment]
@alpe Answered to your reviews and merged as 0 code is a non-issue and has already been introduced to master with the proposal PR a while ago. Feel free to read the link in my comment as it will shine the light on this decision. It's really hard to keep track of these, so that's why I decided to add an extra code comment too.

Maybe we ought to have a wiki doc about errors as consulting PR comments from closed PR is not ideal and it's hard to keep all that in the head.

File: errors/errors.go (hunk 1 above)
Comment (ethanfrey) [reply errors/errors.go]: :point_up:

**Comment by ethanfrey** [comment]
> Maybe we ought to have a wiki doc about errors as consulting PR comments from closed PR is not ideal and it's hard to keep all that in the head.

Well, I would say the best start is godoc in the errors package as a `doc.go`. Or even a `README.md` if preferred.

Keeping doc close to the code helps keep them in sync.
If a wiki, which version does it refer to?

But yeah, please document proper usage and design decisions in the same package

**Pull request merged**